File: trunk/Cantera/src/transport/SolidTransport.cpp
/**
 *
 *  @file SolidTransport.cpp
 */

/* $Author: dggoodwin $
 * $Revision: 1.5 $
 * $Date: 2005/11/10 15:06:33 $
 */

// copyright 2003 California Institute of Technology


// turn off warnings under Windows
#ifdef WIN32
#pragma warning(disable:4786)
#pragma warning(disable:4503)
#endif

#include "../ThermoPhase.h"
#include "SolidTransport.h"

#include "../utilities.h"
#include <iostream>


namespace Cantera {

    //////////////////// class SolidTransport methods //////////////

    SolidTransport::SolidTransport() {}

    void SolidTransport::setParameters(int n, int k, double* p) {
        switch (n) {

            // set the Arrhenius parameters for the diffusion coefficient
            // of species k.
        case 0:
            m_sp.push_back(k);
            m_Adiff.push_back(p[0]);
            m_Ndiff.push_back(p[1]);
            m_Ediff.push_back(p[2]);
            m_nmobile = m_sp.size();
            break;

            // set the thermal conductivity.
        case 1:
            m_lam = p[0];
            break;
        default:
            ;
        }
    }


    /*********************************************************
     *
     *                Public methods
     *
     *********************************************************/


    /**
     * Compute the mobilities of the species from the diffusion coefficients, 
     * using the Einstein relation.
     */
    void SolidTransport::getMobilities(doublereal* mobil) {
        int k;
        getMixDiffCoeffs(mobil);
        doublereal t = m_thermo->temperature();
        int nsp = m_thermo->nSpecies();
        doublereal c1 = ElectronCharge / (Boltzmann * t);
        for (k = 0; k < nsp; k++) {
            mobil[k] *= c1 * fabs(m_thermo->charge(k));
        }
    } 
        

    doublereal SolidTransport::thermalConductivity() {
        return m_lam;
    }


    /**
     * The diffusion coefficients are computed from 
     *
     * \f[
     * D_k = A_k T^{n_k} \exp(-E_k/RT).
     * \f]
     *
     * The diffusion coefficients are only non-zero for species for
     * which parameters have been specified using method
     * setParameters.
     */
    void SolidTransport::getMixDiffCoeffs(doublereal* d) {
        doublereal temp = m_thermo->temperature();
        int nsp = m_thermo->nSpecies();
        int k;
        for (k = 0; k < nsp; k++) d[k] = 0.0;
        for (k = 0; k < m_nmobile; k++) {
            d[m_sp[k]] = 
                m_Adiff[k] * pow(temp, m_Ndiff[k]) * exp(-m_Ediff[k]/temp);
        }
    }

//     void SolidTransport::electricalConductivity() {
//         getMobilities(m_work.begin());
//         int nsp = m_thermo->nSpecies();
//         int k;
//         doublereal sum = 0.0;
//         for (k = 0; k < nsp; n++) {
//             sum += m_thermo->charge(k)*m_thermo->moleFraction(k)*m_work[k];
//         }
//         return sum * m_thermo->molarDensity();
//     }

}

File: cantera/trunk/interfaces/python/Cantera/SolidTransport.py
"""
Transport properties for solids.

This class implements a simple model for the diffusion coefficients and
the thermal conductivity of a solid. The diffusion coefficients have
modified Arrhenius form, and the thermal conductivity is constant.
All parameters are user-specified, not computed from a physical model.

Examples:

    >>> tr = SolidTransport(solid_phase)
    >>> tr.setThermalConductivity(0.5)  # W/m/K
    >>> tr.setDiffCoeff(species = "OxygenIon", A = 2.0, n = 0.0, E = 700.0)

Note that the diffusion coefficient is computed from D = A * T^n *
exp(-E/t) in m^2/s. Diffusion coefficients for unspecified species are
set to zero.

"""

from Cantera.Transport import Transport

class SolidTransport(Transport):
    def __init__(self, phase = None):
        Transport.__init__(self, model = "Solid", phase = phase)

    def setThermalConductivity(self, lamb):
        self.setParameters(1, 0, [lamb, 0.0])

    def setDiffCoeff(self, species = "", A = 0.0, n = 0.0, E = 0.0):
        k = self._phase.speciesIndex(species)
        self.setParameters(0, k, [A, n, E])
